math/src/field/smallprimefield/generic_prime_field_elt.rs

```diff
@@ -0,0 +1,426 @@
+        Self::new(self.value)
+    }
+
+    fn elements_as_bytes(elements: &[Self]) -> &[u8] {
+        let len = elements.len() * Self::ELEMENT_BYTES;
+        let p = elements.as_ptr();
+        unsafe { slice::from_raw_parts(p as *const u8, len) }
+    }
+
+    unsafe fn bytes_as_elements(bytes: &[u8]) -> Result<&[Self], DeserializationError> {
+        if bytes.len() % Self::ELEMENT_BYTES != 0 {
+            return Err(DeserializationError::InvalidValue(format!(
+                "number of bytes ({}) does not divide into whole number of field elements",
+                bytes.len(),
+            )));
+        }
+
+        let p = bytes.as_ptr();
+        let len = bytes.len() / Self::ELEMENT_BYTES;
+
+        if (p as usize) % mem::align_of::<u128>() != 0 {
+            return Err(DeserializationError::InvalidValue(
+                "slice memory alignment is not valid for this field element type".to_string(),
+            ));
+        }
+
+        Ok(slice::from_raw_parts(p as *const Self, len))
+    }
+
+    fn zeroed_vector(n: usize) -> Vec<Self> {
+        // this uses a specialized vector initialization code which requests zero-filled memory
+        // from the OS; here to fulfil trait bounds
+        // We also use u64 to make sure the memory is aligned correctly for our element size.
+        debug_assert_eq!(Self::ELEMENT_BYTES, mem::size_of::<u64>());
+        let result = vec![0u128; n];
+
+        // translate a zero-filled vector of u128s into a vector of base field elements
+        let mut v = core::mem::ManuallyDrop::new(result);
+        let p = v.as_mut_ptr();
+        let len = v.len();
+        let cap = v.capacity();
+        unsafe { Vec::from_raw_parts(p as *mut Self, len, cap) }
+    }
+
+    fn as_base_elements(elements: &[Self]) -> &[Self::BaseField] {
+        elements
+    }
+
+    fn double(self) -> Self {
+        Self::new(self.value + self.value)
+    }
+
+    fn square(self) -> Self {
+        Self::new(self.value * self.value)
+    }
+
+    fn cube(self) -> Self {
+        Self::new(self.value * self.value * self.value)
+    }
+
+    fn exp(self, power: Self::PositiveInteger) -> Self {
+        let mut r = Self::ONE;
+        let mut b = self;
+        let mut p = power;
+
+        let int_zero = Self::PositiveInteger::from(0u32);
+        let int_one = Self::PositiveInteger::from(1u32);
+
+        if p == int_zero {
+            return Self::ONE;
+        } else if b == Self::ZERO {
+            return Self::ZERO;
+        }
+
+        while p > int_zero {
+            if p & int_one == int_one {
+                r *= b;
+            }
+            p >>= int_one;
+            b = b.square();
+        }
+
+        r
+    }
+}
+
+impl<const M: u64, const G: u64> StarkField for GenericPrimeFieldElt<M, G> {
+    type QuadExtension = QuadExtensionA<Self>;
+    const MODULUS: Self::PositiveInteger = M;
+    const MODULUS_BITS: u32 = Self::get_modulus_bits();
+
+    const GENERATOR: Self = Self::new(G);
```

Comment:
I'm not sure this is correct. The generator is supposed to be the primitive element in the field. If it can't be easily inferred from the modulus, maybe pass it as another generic parameter?